torchvision/prototype/datasets/_builtin/cifar.py

```diff
@@ -0,0 +1,227 @@
+import abc
+import functools
+import io
+import pathlib
+import pickle
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union, TypeVar
+
+import numpy as np
+import torch
+from torch.utils.data import IterDataPipe
+from torch.utils.data.datapipes.iter import (
+    Demultiplexer,
+    Filter,
+    Mapper,
+    TarArchiveReader,
+    Shuffler,
+)
+from torchdata.datapipes.iter import KeyZipper
+from torchvision.prototype.datasets.decoder import raw
+from torchvision.prototype.datasets.utils import (
+    Dataset,
+    DatasetConfig,
+    DatasetInfo,
+    HttpResource,
+    OnlineResource,
+    DatasetType,
+)
+from torchvision.prototype.datasets.utils._internal import (
+    create_categories_file,
+    MappingIterator,
+    SequenceIterator,
+    INFINITE_BUFFER_SIZE,
+    image_buffer_from_array,
+    Enumerator,
+)
+
+__all__ = ["Cifar10", "Cifar100"]
+
+HERE = pathlib.Path(__file__).parent
+
+D = TypeVar("D")
+
+
+class _CifarBase(Dataset):
+    @abc.abstractmethod
+    def _is_data_file(self, data: Tuple[str, io.IOBase], *, config: DatasetConfig) -> Optional[int]:
+        pass
+
+    @abc.abstractmethod
+    def _split_data_file(self, data: Tuple[str, Any]) -> Optional[int]:
+        pass
+
+    def _unpickle(self, data: Tuple[str, io.BytesIO]) -> Dict[str, Any]:
+        _, file = data
+        return pickle.load(file, encoding="latin1")
+
+    def _remove_data_dict_key(self, data: Tuple[str, D]) -> D:
+        return data[1]
+
+    def _key_fn(self, data: Tuple[int, Any]) -> int:
+        return data[0]
+
+    def _collate_and_decode(
+        self,
+        data: Tuple[Tuple[int, int], Tuple[int, np.ndarray]],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> Dict[str, Any]:
+        (_, category_idx), (_, image_array_flat) = data
+
+        category = self.categories[category_idx]
+        label = torch.tensor(category_idx)
+
+        image_array = image_array_flat.reshape((3, 32, 32))
+        image: Union[torch.Tensor, io.BytesIO]
+        if decoder is raw:
+            image = torch.from_numpy(image_array)
+        else:
+            image_buffer = image_buffer_from_array(image_array.transpose(1, 2, 0))
+            image = decoder(image_buffer) if decoder else image_buffer
+
+        return dict(label=label, category=category, image=image)
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+        archive_dp = resource_dps[0]
+        archive_dp = TarArchiveReader(archive_dp)
+        archive_dp: IterDataPipe = Filter(archive_dp, functools.partial(self._is_data_file, config=config))
+        archive_dp: IterDataPipe = Mapper(archive_dp, self._unpickle)
```

Comment:
@ejguan any idea how to appease `mypy` here, without slapping `: IterDataPipe` everywhere? Otherwise I'm inclined to blanket ignore `var-annotated` here.